scrapy/dispatch/dispatcher.py

```diff
@@ -0,0 +1,368 @@
+
+
+class Signal(object):
+
+    """
+    Base class for all signals
+
+    Internal attributes:
+
+        receivers
+            { receiverkey (id) : weakref(receiver) }
+    """
+
+    def __init__(self, providing_args=None, use_caching=False):
+        """
+        Create a new signal.
+
+        providing_args
+            A list of the arguments this signal can pass along in a send()
+            call.
+        """
+        self.receivers = []
+        if providing_args is None:
+            providing_args = []
+        self.providing_args = set(providing_args)
+        self.lock = threading.Lock()
+        self.use_caching = use_caching
+        # For convenience we create empty caches even if they are not used.
+        # A note about caching: if use_caching is defined, then for each
+        # distinct sender we cache the receivers that sender has in
+        # 'sender_receivers_cache'. The cache is cleaned when .connect() or
+        # .disconnect() is called and populated on send().
+        self.sender_receivers_cache = weakref.WeakKeyDictionary(
+        ) if use_caching else {}
+        self._dead_receivers = False
+
+    def connect(self, receiver, sender=None, weak=True, dispatch_uid=None):
+        """
+        Connect receiver to sender for signal.
+
+        Arguments:
+
+            receiver
+                A function or an instance method which is to receive signals.
+                Receivers must be hashable objects.
+
+                If weak is True, then receiver must be weak referenceable.
+
+                Receivers must be able to accept keyword arguments.
+
+                If a receiver is connected with a dispatch_uid argument, it
+                will not be added if another receiver was already connected
+                with that dispatch_uid.
+
+            sender
+                The sender to which the receiver should respond. Must either be
+                of type Signal, or None to receive events from any sender.
+
+            weak
+                Whether to use weak references to the receiver. By default, the
+                module will attempt to use weak references to the receiver
+                objects. If this parameter is false, then strong references
+                will be used.
+
+            dispatch_uid
+                An identifier used to uniquely identify a particular instance
+                of a receiver. This will usually be a string, though it may be
+                anything hashable.
+        """
+
+        # If DEBUG is on, check that we got a good receiver
+        settings = get_project_settings()
+        if settings.get('LOG_ENABLED') \
+                and settings.get('LOG_LEVEL') == 'DEBUG':
+            assert callable(receiver), "Signal receivers must be callable."
+            # Check for **kwargs
+            if not func_accepts_kwargs(receiver):
+               raise ValueError(
+                   "Signal receivers must accept keyword arguments(**kwargs).")
```

Comment:
I think we should drop the "if debug" switch here and make this a warning instead of an exception. We have a working backwards-compatible mechanism in place and I see no reason not to use it just because we're using debug level logging.